File: pwdmg_core/passkey_operation.py
from __future__ import annotations

import base64
import binascii
import json
import math
import re
from dataclasses import dataclass
from typing import Any, Dict, Mapping, Optional
from urllib.parse import SplitResult, urlsplit


PASSKEY_OPERATION_PROTOCOL_VERSION = 1
BASE64URL_RE = re.compile(r"^[A-Za-z0-9_-]+$")
ANDROID_ORIGIN_PREFIX = "android:apk-key-hash:"
DOMAIN_LABEL_RE = re.compile(r"^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$")
MAX_REQUEST_BYTES = 64 * 1024
MAX_ORIGIN_BYTES = 2 * 1024
MAX_CHALLENGE_BYTES = 1024
MIN_CHALLENGE_BYTES = 16
MAX_CREDENTIAL_ID_BYTES = 2048
MAX_CREDENTIAL_DESCRIPTORS = 100
MAX_USER_ID_BYTES = 64
MAX_USER_TEXT_BYTES = 512
MAX_RP_TEXT_BYTES = 512
DEFAULT_TIMEOUT_MS = 60_000
MIN_TIMEOUT_MS = 15_000
MAX_TIMEOUT_MS = 120_000
MIN_OPERATION_ID_BYTES = 16
MAX_OPERATION_ID_BYTES = 64
MAX_SAFE_INTEGER = (1 << 53) - 1
MISSING = object()


PasskeyOperationKind = str
PasskeyOperationStatus = str


class PasskeyOperationError(ValueError):
    def __init__(self, code: str):
        super().__init__(code)
        self.code = code


@dataclass(frozen=True)
class PasskeyCreateUser:
    id: str
    name: str
    display_name: str


@dataclass(frozen=True)
class PasskeyOperation:
    protocol_version: int
    kind: PasskeyOperationKind
    origin: str
    rp_id: str
    challenge: str
    client_data_hash: str
    timeout_ms: int
    credential_ids: tuple[str, ...]
    requires_user_verification: bool
    rp_name: Optional[str] = None
    user: Optional[PasskeyCreateUser] = None
    algorithm: Optional[int] = None
    discoverable: bool = False


@dataclass(frozen=True)
class NativePasskeyOperation:
    operation: PasskeyOperation
    operation_id: str
    issued_at: int
    expires_at: int


@dataclass(frozen=True)
class PasskeyOperationOutcome:
    kind: PasskeyOperationKind
    operation_id: str
    status: PasskeyOperationStatus
    platform_response_json: Optional[str] = None


@dataclass(frozen=True)
class PasskeyOperationDiagnostic:
    kind: PasskeyOperationKind
    status: PasskeyOperationStatus
    code: str


def parse_passkey_operation(
    kind: PasskeyOperationKind,
    operation_input: Mapping[str, Any],
) -> PasskeyOperation:
    if kind not in {"create", "get"}:
        _fail("INVALID_REQUEST_JSON")
    if not isinstance(operation_input, Mapping):
        _fail("INVALID_REQUEST_JSON")

    origin, origin_host = _read_trusted_origin(_field(operation_input, "trustedOrigin"))
    client_data_hash = _read_base64url(
        _field(operation_input, "clientDataHash"),
        "INVALID_CLIENT_DATA_HASH",
        32,
        32,
    )
    request = _read_request_json(_field(operation_input, "requestJson"))
    rp_source = _read_object(_field(request, "rp"), "INVALID_RP") if kind == "create" else request
    rp_id = _read_rp_id(_field(rp_source, "id") if kind == "create" else _field(request, "rpId"), origin_host)
    credential_ids = _read_credential_descriptors(
        _field(request, "excludeCredentials") if kind == "create" else _field(request, "allowCredentials")
    )
    base = {
        "protocol_version": PASSKEY_OPERATION_PROTOCOL_VERSION,
        "kind": kind,
        "origin": origin,
        "rp_id": rp_id,
        "challenge": _read_base64url(
            _field(request, "challenge"),
            "INVALID_CHALLENGE",
            MIN_CHALLENGE_BYTES,
            MAX_CHALLENGE_BYTES,
        ),
        "client_data_hash": client_data_hash,
        "timeout_ms": _read_timeout(_field(request, "timeout")),
        "credential_ids": tuple(credential_ids),
        "requires_user_verification": True,
    }

    if kind == "get":
        _read_user_verification(_field(request, "userVerification"))
        return PasskeyOperation(**base)

    rp = _read_object(_field(request, "rp"), "INVALID_RP")
    user = _read_object(_field(request, "user"), "INVALID_USER")
    _read_create_selection(_field(request, "authenticatorSelection"))
    _require_es256(_field(request, "pubKeyCredParams"))
    return PasskeyOperation(
        **base,
        rp_name=_read_wire_text(_field(rp, "name"), "INVALID_RP", MAX_RP_TEXT_BYTES),
        user=PasskeyCreateUser(
            id=_read_base64url(_field(user, "id"), "INVALID_USER", 1, MAX_USER_ID_BYTES),
            name=_read_wire_text(_field(user, "name"), "INVALID_USER", MAX_USER_TEXT_BYTES),
            display_name=_read_wire_text(
                _field(user, "displayName"),
                "INVALID_USER",
                MAX_USER_TEXT_BYTES,
            ),
        ),
        algorithm=-7,
        discoverable=True,
    )


def bind_native_passkey_operation(
    operation: PasskeyOperation,
    operation_id: str,
    issued_at: int,
) -> NativePasskeyOperation:
    canonical_id = _read_base64url(
        operation_id,
        "INVALID_OPERATION_ID",
        MIN_OPERATION_ID_BYTES,
        MAX_OPERATION_ID_BYTES,
    )
    if not _is_safe_integer(issued_at) or issued_at <= 0 or issued_at > MAX_SAFE_INTEGER - operation.timeout_ms:
        _fail("INVALID_OPERATION_TICKET")
    return NativePasskeyOperation(
        operation=operation,
        operation_id=canonical_id,
        issued_at=issued_at,
        expires_at=issued_at + operation.timeout_ms,
    )


def is_passkey_operation_expired(ticket: NativePasskeyOperation, now: int) -> bool:
    if not _is_safe_integer(now) or now <= 0:
        _fail("INVALID_OPERATION_TICKET")
    return now >= ticket.expires_at


def create_succeeded_passkey_operation_outcome(
    ticket: NativePasskeyOperation,
    platform_response_json: str,
) -> PasskeyOperationOutcome:
    _read_platform_response_json(platform_response_json)
    return PasskeyOperationOutcome(
        kind=ticket.operation.kind,
        operation_id=ticket.operation_id,
        status="succeeded",
        platform_response_json=platform_response_json,
    )


def create_cancelled_passkey_operation_outcome(
    ticket: NativePasskeyOperation,
) -> PasskeyOperationOutcome:
    return _create_terminal_outcome(ticket, "cancelled")


def create_rejected_passkey_operation_outcome(
    ticket: NativePasskeyOperation,
) -> PasskeyOperationOutcome:
    return _create_terminal_outcome(ticket, "rejected")


def create_failed_passkey_operation_outcome(
    ticket: NativePasskeyOperation,
) -> PasskeyOperationOutcome:
    return _create_terminal_outcome(ticket, "failed")


def to_passkey_operation_diagnostic(
    outcome: PasskeyOperationOutcome,
) -> PasskeyOperationDiagnostic:
    code: str
    if outcome.status == "succeeded":
        code = "SUCCESS"
    elif outcome.status == "cancelled":
        code = "CANCELLED"
    elif outcome.status == "rejected":
        code = "REJECTED"
    else:
        code = "FAILED"
    return PasskeyOperationDiagnostic(kind=outcome.kind, status=outcome.status, code=code)


def _create_terminal_outcome(
    ticket: NativePasskeyOperation,
    status: str,
) -> PasskeyOperationOutcome:
    return PasskeyOperationOutcome(
        kind=ticket.operation.kind,
        operation_id=ticket.operation_id,
        status=status,
    )


def _read_request_json(value: Any) -> Dict[str, Any]:
    request_json = _read_wire_text(value, "INVALID_REQUEST_JSON", MAX_REQUEST_BYTES)
    try:
        return _read_object(
            json.loads(request_json, parse_constant=_reject_json_constant),
            "INVALID_REQUEST_JSON",
        )
    except (TypeError, ValueError, json.JSONDecodeError) as error:
        if isinstance(error, PasskeyOperationError):
            raise
        raise PasskeyOperationError("INVALID_REQUEST_JSON") from error


def _read_trusted_origin(value: Any) -> tuple[str, Optional[str]]:
    origin = _read_wire_text(value, "INVALID_ORIGIN", MAX_ORIGIN_BYTES)
    if origin.startswith(ANDROID_ORIGIN_PREFIX):
        _read_base64url(origin[len(ANDROID_ORIGIN_PREFIX):], "INVALID_ORIGIN", 32, 32)
        return origin, None
    try:
        split = urlsplit(origin)
        host = split.hostname
        port = split.port
    except ValueError as error:
        raise PasskeyOperationError("INVALID_ORIGIN") from error

    if not _is_canonical_origin(split, origin, host, port):
        _fail("INVALID_ORIGIN")
    if split.scheme not in {"https", "http"} or not _is_webauthn_domain(host):
        _fail("INVALID_ORIGIN")
    if split.scheme == "http" and host != "localhost":
        _fail("INVALID_ORIGIN")
    return origin, host


def _is_canonical_origin(
    split: SplitResult,
    origin: str,
    host: Optional[str],
    port: Optional[int],
) -> bool:
    if not split.scheme or not split.netloc or not host:
        return False
    if split.username is not None or split.password is not None:
        return False
    if split.path or split.query or split.fragment:
        return False
    if split.scheme not in {"https", "http"}:
        return False
    canonical_port = ""
    if port is not None and not (split.scheme == "https" and port == 443) and not (split.scheme == "http" and port == 80):
        canonical_port = f":{port}"
    return origin == f"{split.scheme}://{host}{canonical_port}"


def _read_rp_id(value: Any, origin_host: Optional[str]) -> str:
    if value is MISSING and origin_host is None:
        _fail("INVALID_RP_ID")
    rp_id = origin_host if value is MISSING else _read_wire_text(value, "INVALID_RP_ID", 253).lower()
    if not isinstance(rp_id, str) or not _is_webauthn_domain(rp_id):
        _fail("INVALID_RP_ID")
    return rp_id


def _read_credential_descriptors(value: Any) -> list[str]:
    if value is MISSING:
        return []
    if not isinstance(value, list) or len(value) > MAX_CREDENTIAL_DESCRIPTORS:
        _fail("INVALID_CREDENTIAL_DESCRIPTOR")
    ids: list[str] = []
    seen: set[str] = set()
    for descriptor in value:
        raw = _read_object(descriptor, "INVALID_CREDENTIAL_DESCRIPTOR")
        if _field(raw, "type") != "public-key":
            _fail("INVALID_CREDENTIAL_DESCRIPTOR")
        credential_id = _read_base64url(
            _field(raw, "id"),
            "INVALID_CREDENTIAL_DESCRIPTOR",
            1,
            MAX_CREDENTIAL_ID_BYTES,
        )
        if credential_id in seen:
            _fail("DUPLICATE_CREDENTIAL_ID")
        seen.add(credential_id)
        ids.append(credential_id)
    return ids


def _read_create_selection(value: Any) -> None:
    if value is MISSING:
        return
    selection = _read_object(value, "UNSUPPORTED_RESIDENT_KEY")
    resident_key = _field(selection, "residentKey")
    if resident_key is not MISSING:
        if resident_key not in {"required", "preferred", "discouraged"}:
            _fail("UNSUPPORTED_RESIDENT_KEY")
        if resident_key == "discouraged":
            _fail("UNSUPPORTED_RESIDENT_KEY")
    if "requireResidentKey" in selection and not isinstance(selection["requireResidentKey"], bool):
        _fail("UNSUPPORTED_RESIDENT_KEY")
    authenticator_attachment = _field(selection, "authenticatorAttachment")
    if authenticator_attachment is not MISSING and authenticator_attachment != "platform":
        _fail("UNSUPPORTED_AUTHENTICATOR_ATTACHMENT")
    _read_user_verification(_field(selection, "userVerification"))


def _read_user_verification(value: Any) -> None:
    if value is MISSING:
        return
    if value not in {"required", "preferred", "discouraged"}:
        _fail("INVALID_USER_VERIFICATION")


def _require_es256(value: Any) -> None:
    if not isinstance(value, list) or not value or len(value) > 32:
        _fail("UNSUPPORTED_ALGORITHM")
    for parameter in value:
        if not isinstance(parameter, dict):
            continue
        if parameter.get("type") == "public-key" and _is_integer_value(parameter.get("alg"), -7):
            return
    _fail("UNSUPPORTED_ALGORITHM")


def _read_timeout(value: Any) -> int:
    if value is MISSING:
        return DEFAULT_TIMEOUT_MS
    if not _is_safe_integer(value) or value <= 0:
        _fail("INVALID_TIMEOUT")
    return max(MIN_TIMEOUT_MS, min(MAX_TIMEOUT_MS, int(value)))


def _read_platform_response_json(value: Any) -> None:
    response_json = _read_wire_text(value, "INVALID_PLATFORM_RESPONSE", MAX_REQUEST_BYTES)
    try:
        _read_object(
            json.loads(response_json, parse_constant=_reject_json_constant),
            "INVALID_PLATFORM_RESPONSE",
        )
    except (TypeError, ValueError, json.JSONDecodeError) as error:
        if isinstance(error, PasskeyOperationError):
            raise
        raise PasskeyOperationError("INVALID_PLATFORM_RESPONSE") from error


def _read_object(value: Any, code: str) -> Dict[str, Any]:
    if not isinstance(value, dict):
        _fail(code)
    return value


def _reject_json_constant(_: str) -> None:
    raise ValueError("Non-standard JSON constant")


def _field(value: Mapping[str, Any], key: str) -> Any:
    return value[key] if key in value else MISSING


def _read_wire_text(value: Any, code: str, max_bytes: int) -> str:
    if not isinstance(value, str) or not value or value != value.strip():
        _fail(code)
    if len(value.encode("utf-8")) > max_bytes:
        _fail(code)
    return value


def _read_base64url(value: Any, code: str, min_bytes: int, max_bytes: int) -> str:
    text = _read_wire_text(value, code, max_bytes * 2)
    if not BASE64URL_RE.fullmatch(text):
        _fail(code)
    try:
        decoded = base64.urlsafe_b64decode(text + "=" * (-len(text) % 4))
    except (binascii.Error, ValueError) as error:
        raise PasskeyOperationError(code) from error
    canonical = base64.urlsafe_b64encode(decoded).decode("ascii").rstrip("=")
    if canonical != text or not min_bytes <= len(decoded) <= max_bytes:
        _fail(code)
    return text


def _is_webauthn_domain(value: Optional[str]) -> bool:
    if not value or len(value) > 253 or value.endswith("."):
        return False
    labels = value.split(".")
    if value != "localhost" and len(labels) < 2:
        return False
    return all(DOMAIN_LABEL_RE.fullmatch(label) for label in labels)


def _is_safe_integer(value: Any) -> bool:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return False
    if isinstance(value, float) and not math.isfinite(value):
        return False
    return value == int(value) and 0 < value <= MAX_SAFE_INTEGER


def _is_integer_value(value: Any, expected: int) -> bool:
    return not isinstance(value, bool) and isinstance(value, (int, float)) and value == expected


def _fail(code: str) -> None:
    raise PasskeyOperationError(code)

File: front/src/services/passkeyOperation.ts
export const PASSKEY_OPERATION_PROTOCOL_VERSION = 1

const BASE64URL_RE = /^[A-Za-z0-9_-]+$/
const ANDROID_ORIGIN_PREFIX = 'android:apk-key-hash:'
const DOMAIN_LABEL_RE = /^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$/
const UTF8_ENCODER = new TextEncoder()
const MAX_REQUEST_BYTES = 64 * 1024
const MAX_ORIGIN_BYTES = 2 * 1024
const MAX_CHALLENGE_BYTES = 1024
const MIN_CHALLENGE_BYTES = 16
const MAX_CREDENTIAL_ID_BYTES = 2048
const MAX_CREDENTIAL_DESCRIPTORS = 100
const MAX_USER_ID_BYTES = 64
const MAX_USER_TEXT_BYTES = 512
const MAX_RP_TEXT_BYTES = 512
const DEFAULT_TIMEOUT_MS = 60_000
const MIN_TIMEOUT_MS = 15_000
const MAX_TIMEOUT_MS = 120_000
const MIN_OPERATION_ID_BYTES = 16
const MAX_OPERATION_ID_BYTES = 64
const MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER

export type PasskeyOperationKind = 'create' | 'get'
export type PasskeyOperationErrorCode =
  | 'INVALID_REQUEST_JSON'
  | 'INVALID_ORIGIN'
  | 'INVALID_CLIENT_DATA_HASH'
  | 'INVALID_CHALLENGE'
  | 'INVALID_RP'
  | 'INVALID_RP_ID'
  | 'RP_ID_ORIGIN_MISMATCH'
  | 'INVALID_USER'
  | 'UNSUPPORTED_ALGORITHM'
  | 'INVALID_CREDENTIAL_DESCRIPTOR'
  | 'DUPLICATE_CREDENTIAL_ID'
  | 'INVALID_TIMEOUT'
  | 'UNSUPPORTED_RESIDENT_KEY'
  | 'UNSUPPORTED_AUTHENTICATOR_ATTACHMENT'
  | 'INVALID_USER_VERIFICATION'
  | 'INVALID_OPERATION_ID'
  | 'INVALID_OPERATION_TICKET'
  | 'INVALID_PLATFORM_RESPONSE'

export class PasskeyOperationError extends Error {
  readonly code: PasskeyOperationErrorCode

  constructor(code: PasskeyOperationErrorCode) {
    super(code)
    this.name = 'PasskeyOperationError'
    this.code = code
  }
}

export type PasskeyOperationInput = {
  requestJson: string
  trustedOrigin: string
  clientDataHash: string
}

export type PasskeyCreateUser = {
  id: string
  name: string
  displayName: string
}

type PasskeyOperationBase = {
  protocolVersion: typeof PASSKEY_OPERATION_PROTOCOL_VERSION
  kind: PasskeyOperationKind
  origin: string
  rpId: string
  challenge: string
  clientDataHash: string
  timeoutMs: number
  credentialIds: string[]
  requiresUserVerification: true
}

export type PasskeyCreateOperation = PasskeyOperationBase & {
  kind: 'create'
  rpName: string
  user: PasskeyCreateUser
  algorithm: -7
  discoverable: true
}

export type PasskeyGetOperation = PasskeyOperationBase & {
  kind: 'get'
}

export type PasskeyOperation = PasskeyCreateOperation | PasskeyGetOperation

export type NativePasskeyOperation = {
  operation: PasskeyOperation
  operationId: string
  issuedAt: number
  expiresAt: number
}

export type PasskeyOperationOutcome =
  | {
    kind: PasskeyOperationKind
    operationId: string
    status: 'succeeded'
    platformResponseJson: string
  }
  | {
    kind: PasskeyOperationKind
    operationId: string
    status: 'cancelled' | 'rejected' | 'failed'
  }

export type PasskeyOperationDiagnostic = {
  kind: PasskeyOperationKind
  status: PasskeyOperationOutcome['status']
  code: 'SUCCESS' | 'CANCELLED' | 'REJECTED' | 'FAILED'
}

export function parsePasskeyOperation(
  kind: PasskeyOperationKind,
  input: PasskeyOperationInput
): PasskeyOperation {
  const origin = readTrustedOrigin(input.trustedOrigin)
  const clientDataHash = readBase64Url(
    input.clientDataHash,
    'INVALID_CLIENT_DATA_HASH',
    32,
    32
  )
  const request = readRequestJson(input.requestJson)
  const base = {
    protocolVersion: 1 as const,
    origin: origin.value,
    rpId: readRpId(kind === 'create' ? readObject(request.rp, 'INVALID_RP').id : request.rpId, origin.host),
    challenge: readBase64Url(request.challenge, 'INVALID_CHALLENGE', MIN_CHALLENGE_BYTES, MAX_CHALLENGE_BYTES),
    clientDataHash,
    timeoutMs: readTimeout(request.timeout),
    credentialIds: readCredentialDescriptors(
      kind === 'create' ? request.excludeCredentials : request.allowCredentials
    ),
    requiresUserVerification: true as const
  }

  if (kind === 'get') {
    readUserVerification(request.userVerification)
    return { ...base, kind: 'get' }
  }

  const rp = readObject(request.rp, 'INVALID_RP')
  const user = readObject(request.user, 'INVALID_USER')
  readCreateSelection(request.authenticatorSelection)
  requireEs256(request.pubKeyCredParams)
  return {
    ...base,
    kind: 'create',
    rpName: readWireText(rp.name, 'INVALID_RP', MAX_RP_TEXT_BYTES),
    user: {
      id: readBase64Url(user.id, 'INVALID_USER', 1, MAX_USER_ID_BYTES),
      name: readWireText(user.name, 'INVALID_USER', MAX_USER_TEXT_BYTES),
      displayName: readWireText(user.displayName, 'INVALID_USER', MAX_USER_TEXT_BYTES)
    },
    algorithm: -7,
    discoverable: true
  }
}

export function bindNativePasskeyOperation(
  operation: PasskeyOperation,
  operationId: string,
  issuedAt: number
): NativePasskeyOperation {
  const canonicalId = readBase64Url(
    operationId,
    'INVALID_OPERATION_ID',
    MIN_OPERATION_ID_BYTES,
    MAX_OPERATION_ID_BYTES
  )
  if (!Number.isSafeInteger(issuedAt) || issuedAt <= 0 || issuedAt > MAX_SAFE_INTEGER - operation.timeoutMs) {
    fail('INVALID_OPERATION_TICKET')
  }
  return {
    operation,
    operationId: canonicalId,
    issuedAt,
    expiresAt: issuedAt + operation.timeoutMs
  }
}

export function isPasskeyOperationExpired(ticket: NativePasskeyOperation, now: number): boolean {
  if (!Number.isSafeInteger(now) || now <= 0) fail('INVALID_OPERATION_TICKET')
  return now >= ticket.expiresAt
}

export function createSucceededPasskeyOperationOutcome(
  ticket: NativePasskeyOperation,
  platformResponseJson: string
): PasskeyOperationOutcome {
  readPlatformResponseJson(platformResponseJson)
  return {
    kind: ticket.operation.kind,
    operationId: ticket.operationId,
    status: 'succeeded',
    platformResponseJson
  }
}

export function createCancelledPasskeyOperationOutcome(ticket: NativePasskeyOperation): PasskeyOperationOutcome {
  return createTerminalPasskeyOperationOutcome(ticket, 'cancelled')
}

export function createRejectedPasskeyOperationOutcome(ticket: NativePasskeyOperation): PasskeyOperationOutcome {
  return createTerminalPasskeyOperationOutcome(ticket, 'rejected')
}

export function createFailedPasskeyOperationOutcome(ticket: NativePasskeyOperation): PasskeyOperationOutcome {
  return createTerminalPasskeyOperationOutcome(ticket, 'failed')
}

export function toPasskeyOperationDiagnostic(outcome: PasskeyOperationOutcome): PasskeyOperationDiagnostic {
  const code = outcome.status === 'succeeded'
    ? 'SUCCESS'
    : outcome.status === 'cancelled'
      ? 'CANCELLED'
      : outcome.status === 'rejected'
        ? 'REJECTED'
        : 'FAILED'
  return {
    kind: outcome.kind,
    status: outcome.status,
    code
  }
}

function createTerminalPasskeyOperationOutcome(
  ticket: NativePasskeyOperation,
  status: 'cancelled' | 'rejected' | 'failed'
): PasskeyOperationOutcome {
  return {
    kind: ticket.operation.kind,
    operationId: ticket.operationId,
    status
  }
}

function readRequestJson(value: unknown): Record<string, unknown> {
  const requestJson = readWireText(value, 'INVALID_REQUEST_JSON', MAX_REQUEST_BYTES)
  try {
    return readObject(JSON.parse(requestJson), 'INVALID_REQUEST_JSON')
  } catch (error) {
    if (error instanceof PasskeyOperationError) throw error
    fail('INVALID_REQUEST_JSON')
  }
}

function readTrustedOrigin(value: unknown): { value: string, host: string | null } {
  const origin = readWireText(value, 'INVALID_ORIGIN', MAX_ORIGIN_BYTES)
  if (origin.startsWith(ANDROID_ORIGIN_PREFIX)) {
    readBase64Url(origin.slice(ANDROID_ORIGIN_PREFIX.length), 'INVALID_ORIGIN', 32, 32)
    return { value: origin, host: null }
  }
  let url: URL
  try {
    url = new URL(origin)
  } catch {
    fail('INVALID_ORIGIN')
  }
  if (url.origin === 'null' || origin !== url.origin) fail('INVALID_ORIGIN')
  const scheme = url.protocol
  const host = url.hostname
  if ((scheme !== 'https:' && scheme !== 'http:') || !isWebAuthnDomain(host)) {
    fail('INVALID_ORIGIN')
  }
  if (scheme === 'http:' && host !== 'localhost') fail('INVALID_ORIGIN')
  return { value: origin, host }
}

function readRpId(value: unknown, originHost: string | null): string {
  if (value === undefined && !originHost) fail('INVALID_RP_ID')
  const rpId = value === undefined
    ? originHost as string
    : readWireText(value, 'INVALID_RP_ID', 253).toLowerCase()
  if (!isWebAuthnDomain(rpId)) fail('INVALID_RP_ID')
  return rpId
}

function readCredentialDescriptors(value: unknown): string[] {
  if (value === undefined) return []
  if (!Array.isArray(value) || value.length > MAX_CREDENTIAL_DESCRIPTORS) {
    fail('INVALID_CREDENTIAL_DESCRIPTOR')
  }
  const ids: string[] = []
  const seen = new Set<string>()
  for (const descriptor of value) {
    const raw = readObject(descriptor, 'INVALID_CREDENTIAL_DESCRIPTOR')
    if (raw.type !== 'public-key') fail('INVALID_CREDENTIAL_DESCRIPTOR')
    const id = readBase64Url(
      raw.id,
      'INVALID_CREDENTIAL_DESCRIPTOR',
      1,
      MAX_CREDENTIAL_ID_BYTES
    )
    if (seen.has(id)) fail('DUPLICATE_CREDENTIAL_ID')
    seen.add(id)
    ids.push(id)
  }
  return ids
}

function readCreateSelection(value: unknown): void {
  if (value === undefined) return
  const selection = readObject(value, 'UNSUPPORTED_RESIDENT_KEY')
  const residentKey = selection.residentKey
  if (residentKey !== undefined) {
    if (residentKey !== 'required' && residentKey !== 'preferred' && residentKey !== 'discouraged') {
      fail('UNSUPPORTED_RESIDENT_KEY')
    }
    if (residentKey === 'discouraged') fail('UNSUPPORTED_RESIDENT_KEY')
  }
  if (selection.requireResidentKey !== undefined && typeof selection.requireResidentKey !== 'boolean') {
    fail('UNSUPPORTED_RESIDENT_KEY')
  }
  if (selection.authenticatorAttachment !== undefined && selection.authenticatorAttachment !== 'platform') {
    fail('UNSUPPORTED_AUTHENTICATOR_ATTACHMENT')
  }
  readUserVerification(selection.userVerification)
}

function readUserVerification(value: unknown): void {
  if (value === undefined) return
  if (value !== 'required' && value !== 'preferred' && value !== 'discouraged') {
    fail('INVALID_USER_VERIFICATION')
  }
}

function requireEs256(value: unknown): void {
  if (!Array.isArray(value) || value.length === 0 || value.length > 32) {
    fail('UNSUPPORTED_ALGORITHM')
  }
  for (const parameter of value) {
    if (!isObject(parameter)) continue
    if (parameter.type === 'public-key' && parameter.alg === -7) return
  }
  fail('UNSUPPORTED_ALGORITHM')
}

function readTimeout(value: unknown): number {
  if (value === undefined) return DEFAULT_TIMEOUT_MS
  if (typeof value !== 'number' || !Number.isSafeInteger(value) || value <= 0) {
    fail('INVALID_TIMEOUT')
  }
  return Math.max(MIN_TIMEOUT_MS, Math.min(MAX_TIMEOUT_MS, value))
}

function readPlatformResponseJson(value: unknown): void {
  const responseJson = readWireText(value, 'INVALID_PLATFORM_RESPONSE', MAX_REQUEST_BYTES)
  try {
    readObject(JSON.parse(responseJson), 'INVALID_PLATFORM_RESPONSE')
  } catch (error) {
    if (error instanceof PasskeyOperationError) throw error
    fail('INVALID_PLATFORM_RESPONSE')
  }
}

function readObject(value: unknown, code: PasskeyOperationErrorCode): Record<string, unknown> {
  if (!isObject(value)) fail(code)
  return value
}

function isObject(value: unknown): value is Record<string, unknown> {
  return Boolean(value) && typeof value === 'object' && !Array.isArray(value)
}

function readWireText(value: unknown, code: PasskeyOperationErrorCode, maxBytes: number): string {
  if (typeof value !== 'string' || !value || value !== value.trim()) fail(code)
  if (UTF8_ENCODER.encode(value).byteLength > maxBytes) fail(code)
  return value
}

function readBase64Url(
  value: unknown,
  code: PasskeyOperationErrorCode,
  minBytes: number,
  maxBytes: number
): string {
  const text = readWireText(value, code, maxBytes * 2)
  if (!BASE64URL_RE.test(text)) fail(code)
  let decoded: string
  try {
    const padding = '='.repeat((4 - (text.length % 4)) % 4)
    decoded = atob(text.replace(/-/g, '+').replace(/_/g, '/') + padding)
  } catch {
    fail(code)
  }
  const canonical = btoa(decoded).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '')
  if (canonical !== text || decoded.length < minBytes || decoded.length > maxBytes) fail(code)
  return text
}

function isWebAuthnDomain(value: string): boolean {
  if (!value || value.length > 253 || value.endsWith('.')) return false
  const labels = value.split('.')
  if (value !== 'localhost' && labels.length < 2) return false
  return labels.every((label) => DOMAIN_LABEL_RE.test(label))
}

function fail(code: PasskeyOperationErrorCode): never {
  throw new PasskeyOperationError(code)
}
